include/rosparam_shortcuts/rosparam_shortcuts.h

```diff
@@ -106,6 +110,249 @@ bool convertDoublesToEigen(const std::vector<double>& values, Eigen::Isometry3d&
+  : node_(node), logger_(logger) {}
+
+template <typename ParameterT>
+void NodeParameters::declareAndGet(const std::string& name, const ParameterT& default_value,
+                                   const rcl_interfaces::msg::ParameterDescriptor& parameter_descriptor,
+                                   const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name, default_value, parameter_descriptor);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+template <typename ParameterT>
+void NodeParameters::declareAndGet(const std::string& name, const ParameterT& default_value,
+                                   const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name, default_value);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+template <typename ParameterT>
+void NodeParameters::declareAndGet(const std::string& name, const ParamCallback& callback) {
+	std::lock_guard<std::mutex> guard(mutex_);
+	// set our internal state for validate
+	validate_.emplace(name, callback);
+
+	// declare and get through ros
+	if (!node_->has_parameter(name)) {
+		node_->declare_parameter(name);
+	}
+
+	// set our internal state of the parameter
+	parameters_.emplace(name, node_->get_parameter(name));
+}
+
+void NodeParameters::registerRosCallbackReconfigure() {
+	on_set_callback_handle_ = node_->add_on_set_parameters_callback(
+	    [this](const std::vector<rclcpp::Parameter>& parameters) { return setParametersCallback(parameters); });
+}
+
+void NodeParameters::registerUpdateCallback(const std::function<void()>& callback) const {
+	std::lock_guard<std::mutex> guard(mutex_);
+	changed_notify_callback_.push_back(callback);
+}
+
+bool NodeParameters::get(const std::string& name, rclcpp::Parameter& parameter) const {
+	std::lock_guard<std::mutex> guard(mutex_);
+	auto internal_param = parameters_.find(name);
+	if (internal_param != parameters_.end()) {
+		parameter = parameters_.at(name);
+		return true;
+	}
+
+	return false;
+}
+
+rclcpp::Parameter NodeParameters::get(const std::string& name) const {
+	std::lock_guard<std::mutex> guard(mutex_);
+	auto internal_param = parameters_.find(name);
+	if (internal_param != parameters_.end()) {
+		return parameters_.at(name);
+	} else {
+		RCLCPP_ERROR_STREAM(logger_, "Parameter: " << name + " not found.");
+	}
+
+	return rclcpp::Parameter();
+}
+
+rcl_interfaces::msg::SetParametersResult
+NodeParameters::setParametersCallback(const std::vector<rclcpp::Parameter>& parameters) {
```

Comment:
Could this be moved into a C++ file (along with the rest of the implementations in here) and out of a header file?  This way (as a header only library) each time this is included it has to be compiled.